sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java

```diff
@@ -177,35 +176,48 @@ private Row buildResponse(Map<String, Long> apeTimings, Map<String, Map<String,
-            AtomicInteger counter = new AtomicInteger(nodeIds.size());
-
-            for (String nodeId : nodeIds) {
-                collectProfileOperation.collect(nodeId)
-                    .whenComplete((map, throwable) -> {
-                        if (throwable == null) {
-                            mergedMap.put(nodeId, map);
-                            if (counter.decrementAndGet() == 0) {
-                                resultFuture.complete(mergedMap);
-                            }
-                        } else {
-                            resultFuture.completeExceptionally(throwable);
-                        }
-                    });
-            }
-            return resultFuture;
-        } else {
-            // Collect from local JobExecutionContext
-            return executor
+        CompletableFuture<Map<String, Map<String, Long>>> resultFuture = new CompletableFuture<>();
+        TransportCollectProfileOperation collectProfileOperation = getTransportCollectProfileOperation(executor, jobId);
+
+        ConcurrentHashMap<String, Map<String, Long>> mergedMap = new ConcurrentHashMap<>(nodeIds.size());
+        boolean needsCollectLocal = !nodeIds.contains(executor.localNodeId());
+
+        AtomicInteger counter = new AtomicInteger(nodeIds.size());
+        if (needsCollectLocal) {
+            counter.incrementAndGet();
+        }
+
+        // Collect from remote JobExecutionContexts
+        for (String nodeId : nodeIds) {
+            collectProfileOperation.collect(nodeId)
+                .whenComplete(mergeResultsAndCompleteFuture(resultFuture, mergedMap, counter, nodeId));
+        }
+
+        // Also collect from local JobExecutionContext if needed
+        if (needsCollectLocal) {
+            executor
                 .transportActionProvider()
                 .transportCollectProfileNodeAction()
                 .collectExecutionTimesAndFinishContext(jobId)
-                .thenApply(timings -> Collections.singletonMap(executor.localNodeId(), timings));
+                .whenComplete(mergeResultsAndCompleteFuture(resultFuture, mergedMap, counter, executor.localNodeId()));
         }
+        return resultFuture;
+    }
+
+    private BiConsumer<Map<String, Long>, Throwable> mergeResultsAndCompleteFuture(CompletableFuture<Map<String, Map<String, Long>>> resultFuture,
+                                                                                   ConcurrentHashMap<String, Map<String, Long>> mergedMap,
+                                                                                   AtomicInteger counter,
+                                                                                   String nodeId) {
+        return (map, throwable) -> {
+            if (throwable == null) {
+                mergedMap.put(nodeId, map);
+                if (counter.decrementAndGet() == 0) {
+                    resultFuture.complete(mergedMap);
```

Comment:
Where does it close the local collect context? Does that happen if we complete the future here?

Reply:
the context (both remote or local) is closed whenever the `profilingFuture` is completed
https://github.com/crate/crate/blob/920c4dc4a7488547e93e91b34ef408214477d49d/sql/src/main/java/io/crate/execution/jobs/JobExecutionContext.java#L280